Move mps_linear forward to use MPS kernels directly instead of MPSGraph

jhavukainen opened the pull request:
This PR moves `mps_linear` to use MPSNDArrays and call into the MPS kernel directly instead of going through MPSGraph. It also adds a caching mechanism for reusing MPS kernels as there is also a small overhead attached to creating the kernel object.

The impact of the improvement is relatively more significant for small input kernels where the MPSGraph overhead represents a larger portion of the overall execution time of the operation but the speedup shows for both small and large input sizes as expected.

`mps_linear` before the changes:
```
input shapes: f32:[1,1,20], f32:[1,20]
torch.linear time: <torch.utils.benchmark.utils.common.Measurement object at 0x109d67110>
func(*args, **kwargs)
  Median: 199.29 us
  IQR:    9.56 us (196.71 to 206.27)
  979 measurements, 1 runs per measurement, 1 thread

input shapes: f32:[1,1,5120], f32:[13284,5120]
torch.linear time: <torch.utils.benchmark.utils.common.Measurement object at 0x1063b4510>
func(*args, **kwargs)
  Median: 979.29 us
  IQR:    25.29 us (964.83 to 990.13)
  205 measurements, 1 runs per measurement, 1 thread
```

`mps_linear` after the changes:
```
input shapes: f32:[1,1,20], f32:[1,20]
torch.linear time: <torch.utils.benchmark.utils.common.Measurement object at 0x10693a190>
func(*args, **kwargs)
  Median: 176.08 us
  IQR:    15.02 us (172.42 to 187.44)
  1103 measurements, 1 runs per measurement, 1 thread

input shapes: f32:[1,1,5120], f32:[13284,5120]
torch.linear time: <torch.utils.benchmark.utils.common.Measurement object at 0x10d524dd0>
func(*args, **kwargs)
  Median: 952.56 us
  IQR:    15.63 us (945.47 to 961.10)
  210 measurements, 1 runs per measurement, 1 thread
```

cc @kulinseth @albanD @malfet @DenisVieriu97

jhavukainen committed:
Adding a direct MPS kernel path to linear op and MPS kernel caching mechanism for improved perf.

malfet commented on aten/src/ATen/native/mps/OperationUtils.h:
```diff
@@ -160,6 +160,22 @@ string get_mem_format_string(c10::MemoryFormat memory_format);
 using MPSCacheKey = uint64_t;
 
+struct MPSCachedKernel {
+  MPSCachedKernel(NSObject* object) : _object([object retain]) {}
```
Please either add a copy constructor or delete it, otherwise it'll be very prone to errors

jhavukainen replied:
Deleted it

jhavukainen committed:
Delete copy constructor and assignment from MPSCachedKernel. Include
MPSSequoiaOps header to Linear. Add check for macos version to
use correct code path to avoid breaking previous OS.

jhavukainen committed:
Use the proper way of addressing strides when creating MPSNDArray

jhavukainen committed:
Get encoder after call to endKernelCoalescing

kulinseth reviewed (approved):
Looks good.

malfet commented on aten/src/ATen/native/mps/operations/Linear.mm:
```diff
@@ -54,66 +113,70 @@ Tensor _mps_linear(const Tensor& input, const Tensor& weight_arg, const std::opt
-      auto inputFlattened = doReshape ? [mpsGraph flatten2DTensor:inputTensor axis:-1 name:nil] : inputTensor;
-      auto outputTensor = [mpsGraph matrixMultiplicationWithPrimaryTensor:inputFlattened
-                                                          secondaryTensor:weightTransposeTensor
+  bool is_macos_15_or_newer = is_macos_13_or_newer(MacOSVersion::MACOS_VER_15_0_PLUS);
+  if (is_macos_15_or_newer) {
+    _mps_linear_nograph(input, weight, bias, output);
+  } else {
```
IMO it would be nice to just replace it with `if() { return;}` (Doing it now)

malfet committed:
Return early

malfet committed:
Fix lint

malfet reviewed (approved):
LGTM, though feels a bit like too much boilerplate code, but will look into it later

malfet committed:
Fix typo

malfet commented:
@pytorchbot merge -f "Lint + MPS are green"
